horovod/common/response_cache.cc

```diff
@@ -0,0 +1,223 @@
+    return CacheState::MISS;
+  }
+}
+
+void ResponseCache::put_(const Response& response, TensorParams& params) {
+  uint32_t cache_bit;
+  auto cache_state = this->cached(response, params);
+
+  // Disallow caching name-conflicted responses here. Invalid cache entries
+  // must be removed prior to caching new entries.
+  if (cache_state == CacheState::INVALID) {
+    throw std::logic_error(
+        "Trying to overwrite cached response with existing name. "
+        "This is not allowed.");
+  }
+
+  if (this->cached(response, params) == CacheState::HIT) {
+    cache_bit = tensor_name_to_bit_[response.tensor_names()[0]];
+    auto it = cache_iters_[cache_bit];
+    cache_.push_front(std::move(*it));
+    cache_.erase(it);
+  } else if (cache_.size() == capacity_) {
+    auto& entry = cache_.back().first;
+    cache_bit = tensor_name_to_bit_[entry.tensor_names()[0]];
+    tensor_name_to_bit_.erase(entry.tensor_names()[0]);
+    cache_.pop_back();
+    cache_.push_front(std::make_pair(response, std::move(params)));
+  } else {
+    cache_bit = cache_iters_.size();
+    cache_iters_.resize(cache_bit + 1);
+    cache_.push_front(std::make_pair(response, std::move(params)));
+  }
+
+  cache_iters_[cache_bit] = cache_.begin();
+  tensor_name_to_bit_[response.tensor_names()[0]] = cache_bit;
+
+  bits_outdated_ = true;
+}
+
+void ResponseCache::put(const Response& response, const TensorTable& tensor_table) {
+  if (capacity_ == 0) {
+    return;
+  }
+
+  // If response is fused, split back into individual responses
+  if (response.tensor_names().size() > 1) {
+    for (auto& name : response.tensor_names()) {
+      Response new_response;
+      new_response.add_tensor_name(name);
+      new_response.set_response_type(response.response_type());
+      new_response.set_devices(response.devices());
+      new_response.set_tensor_sizes(response.tensor_sizes());
+
+      // Populate tensor parameters from tensor_table entry
+      auto& tensor_entry = tensor_table.at(name);
+      TensorParams params;
+      params.device = tensor_entry.device;
+      params.dtype = tensor_entry.tensor->dtype();
+      params.shape = tensor_entry.tensor->shape().to_vector();
+
+      this->put_(new_response, params);
+    }
+  } else {
+    auto& tensor_entry = tensor_table.at(response.tensor_names()[0]);
+    TensorParams params;
+    params.device = tensor_entry.device;
+    params.dtype = tensor_entry.tensor->dtype();
+    params.shape = tensor_entry.tensor->shape().to_vector();
+
+    this->put_(response, params);
+  }
+}
+
+const Response& ResponseCache::get_response(const Request& message) {
+  assert(this->cached(message));
+  uint32_t cache_bit = tensor_name_to_bit_[message.tensor_name()];
+  auto it = cache_iters_[cache_bit];
+  cache_.push_front(std::move(*it));
+  cache_.erase(it);
+  cache_iters_[cache_bit] = cache_.begin();
+  bits_outdated_ = true;
+  return cache_.front().first;
+}
+
+const Response& ResponseCache::get_response(uint32_t cache_bit) {
```

Comment:
Let's add more comments, e.g. how this method pulls the cached response to the front of the cache?

Reply:
I added some more comments to the various methods in `response_cache.cc`. Please take a look and let me know if anything jumps out at you that we can simplify.